subsys/emds/emds_flash.c

```diff
@@ -283,22 +339,23 @@ static int entry_wrt(struct emds_fs *fs, uint16_t id, const void *data, size_t l
 static enum ate_type ate_check(struct emds_fs *fs, uint32_t addr, struct emds_ate *entry)
 {
 	uint8_t cmp_buf[fs->ate_size];
-	int rc = flash_read(fs->flash_dev, addr, entry, fs->ate_size);
+	int rc = flash_read(fs->flash_dev, addr, cmp_buf, fs->ate_size);
 
 	if (rc) {
 		return ATE_TYPE_UNKNOWN;
 	}
 
-	memset(cmp_buf, 0, sizeof(cmp_buf));
-	if (!memcmp(entry, cmp_buf, fs->ate_size)) {
+	memset(entry, 0, sizeof(struct emds_ate));
+	if (!memcmp(entry, cmp_buf, sizeof(struct emds_ate))) {
 		return ATE_TYPE_INVALIDATED;
 	}
 
-	memset(cmp_buf, fs->flash_params->erase_value, sizeof(cmp_buf));
-	if (!memcmp(entry, cmp_buf, fs->ate_size)) {
+	memset(entry, fs->flash_params->erase_value, sizeof(struct emds_ate));
```

Comment:
Is there a reason why you flipped the usage of `entry` and `cmp_buf` around? As far as I can see, the behavior of this function is identical before and after your changes, but now you need the final `memcpy` on line 358 since you don't read directly into `entry`. If there is a reason for this that I can't see, then I'd suggest to rename `cmp_buf` to something like `read_entry` or `stored_entry` or `flash_data`, since it's no longer a compare buffer.

Reply:
Yes, there is. `sizeof(struct emds_ate) < fs->ate_size` for rram. The existing implementation corrupted stack memory that caused hardfaults in unexpected places.
As I said before, emds has been based on assumption the ATE structure is bigger than writing granularity of persistent memory. That's not the truth for flash emulator on rram. 

In general, my implementation didn't add more data copying. Yes, it added one more on line 358 but removed it on line 368.

Reply:
Right, I missed the size being different. Then I guess my only suggestion would be to rename `cmp_buf` which is a bit confusingly named now after you flipped things around

Reply:
I also thought this looked strange at first, but from a functional view it looks fine. I would claim that it is a bit more vulnerable for introducing bugs if this is is changed in the future, but im ok with it.